### PR TITLE
Add a "fees" section to the ERC4626 guide

### LINKED ISSUE #3814: Support fees in ERC4626 vaults
EIP-4626 says "Fees can exists for deposits, yield, AUM, withdrawals, or anything else prescribed by the Vault."

Fees for deposits and withdrawals can be hard to implement correctly and compliantly given that there are requirements on how functions interact with fees. For example:

> previewDeposit MUST be inclusive of deposit fees

Our current code may not be easily adaptable to add fees with inheritance in a compliant way. This is something we need to review.

We didn't include fees by default because there are many things that can have fees added and we didn't want to add the complexity to handle all of them.

It may be worth adding support for just deposit and withdrawal fees if those are particularly hard to implement.

If we don't add support, we should at least document how users can add the fees to a vault.

### PR DESCRIPTION
**Pull request by Amxx**
Fixes #3814
Fixes LIB-646

#### PR Checklist

- [x] Documentation
- [ ] Changeset entry (run `npx changeset add`)
  - [x] None for doc updates

### CODE REVIEW

File: docs/modules/ROOT/pages/erc4626.adoc [hunk 1]
```diff
@@ -191,3 +191,99 @@ stem:[\delta = 3], stem:[a_0 = 100], stem:[a_1 = 10^5]
...
+
+For example, if calling `deposit(100, receiver)`, the caller should deposit exactly 100 underlying tokens, including fees, and the receiver should receive a number of shares that matches the value returned by `previewDeposit(100)`. Similarly, `previewMint` should account for the fees that the user will have to pay on top of share's cost.
+
+As for the `Deposit` event, we believe it should include the number of assets paid for by the user, including the fees.
+
+On the other hand, when withdrawing assets, the number given by the user should correspond to what he receives. Any fees should be added to the quote (in shares) performed by `previewWithdraw`.
+
+The `Withdraw` event should include the number of shares the users burns (including fees) and the number of assets the user actually receives (after fees are deducted).
+
+The consequence of this design is that both the `Deposit` and `Withdraw` events will describe two different prices. The spread between the "Buy-in" price and the "Exit" price correspong to the fees taken by the vault.
```
Comment (Amxx) [inline docs/modules/ROOT/pages/erc4626.adoc:209]:
```suggestion
The consequence of this design is that both the `Deposit` and `Withdraw` events will describe two exchange rates. The spread between the "Buy-in" price and the "Exit" prices correspond to the fees taken by the vault.
```

File: docs/modules/ROOT/pages/erc4626.adoc [hunk 2]
```diff
@@ -191,3 +191,99 @@ stem:[\delta = 3], stem:[a_0 = 100], stem:[a_1 = 10^5]
...
+
+For example, if calling `deposit(100, receiver)`, the caller should deposit exactly 100 underlying tokens, including fees, and the receiver should receive a number of shares that matches the value returned by `previewDeposit(100)`. Similarly, `previewMint` should account for the fees that the user will have to pay on top of share's cost.
+
+As for the `Deposit` event, we believe it should include the number of assets paid for by the user, including the fees.
+
+On the other hand, when withdrawing assets, the number given by the user should correspond to what he receives. Any fees should be added to the quote (in shares) performed by `previewWithdraw`.
+
+The `Withdraw` event should include the number of shares the users burns (including fees) and the number of assets the user actually receives (after fees are deducted).
+
+The consequence of this design is that both the `Deposit` and `Withdraw` events will describe two exchange rates. The spread between the "Buy-in" price and the "Exit" prices correspond to the fees taken by the vault.
```
Comment (Amxx) [inline docs/modules/ROOT/pages/erc4626.adoc:209]:
```suggestion
The consequence of this design is that both the `Deposit` and `Withdraw` events will describe two exchange rates. The spread between the "Buy-in" and the "Exit" prices correspond to the fees taken by the vault.
```

File: docs/modules/ROOT/pages/erc4626.adoc [hunk 3]
```diff
@@ -191,3 +191,99 @@ stem:[\delta = 3], stem:[a_0 = 100], stem:[a_1 = 10^5]
...
+
+    // Vault behavior override
+    function previewDeposit(uint256 assets) public view virtual override returns (uint256) {
+        uint256 fee = _feeOnTotal(assets, _entryFeeBasePoint());
+        return super.previewDeposit(assets - fee);
+    }
+
+    function previewMint(uint256 shares) public view virtual override returns (uint256) {
+        uint256 assets = super.previewMint(shares);
+        return assets + _feeOnTotal(assets, _entryFeeBasePoint());
```
Comment (Amxx) [inline docs/modules/ROOT/pages/erc4626.adoc:235]:
```suggestion
        return assets + _feeOnRaw(assets, _entryFeeBasePoint());
```

**Comment by Amxx** [comment]
If we were able to import the solidity source code from a file then we could:
- add `ERC4626Fees.sol` to the codebase (`/contracts/mocks/docs` ?)
- import the code into the docs
- add tests.

The code and tests is already available in [this branch](https://github.com/Amxx/openzeppelin-contracts/tree/feature/ERC4626/fee)

File: docs/modules/ROOT/pages/erc4626.adoc [hunk 4]
```diff
@@ -191,3 +191,99 @@ stem:[\delta = 3], stem:[a_0 = 100], stem:[a_1 = 10^5]
...
+
+
+[[fees]]
+== Custom behavior: Adding fees to the vault
+
+In an ERC4626 vaults, fees can be captured during the deposit/mint and/or during the withdraw/redeem steps. In both cases it is essential to remain compliant with the ERC4626 requirements with regard to the preview functions.
+
+For example, if calling `deposit(100, receiver)`, the caller should deposit exactly 100 underlying tokens, including fees, and the receiver should receive a number of shares that matches the value returned by `previewDeposit(100)`. Similarly, `previewMint` should account for the fees that the user will have to pay on top of share's cost.
+
+As for the `Deposit` event, we believe it should include the number of assets paid for by the user, including the fees.
```
Comment (frangio) [inline docs/modules/ROOT/pages/erc4626.adoc:203]:
```suggestion
As for the `Deposit` event, while this is less clear in the EIP spec itself, there seems to be consensus that it should include the number of assets paid for by the user, including the fees.
```

File: docs/modules/ROOT/pages/erc4626.adoc [hunk 5]
```diff
@@ -191,3 +191,99 @@ stem:[\delta = 3], stem:[a_0 = 100], stem:[a_1 = 10^5]
...
+
+On the other hand, when withdrawing assets, the number given by the user should correspond to what he receives. Any fees should be added to the quote (in shares) performed by `previewWithdraw`.
+
+The `Withdraw` event should include the number of shares the users burns (including fees) and the number of assets the user actually receives (after fees are deducted).
```
Comment (frangio) [inline docs/modules/ROOT/pages/erc4626.adoc:207]:
```suggestion
The `Withdraw` event should include the number of shares the user burns (including fees) and the number of assets the user actually receives (after fees are deducted).
```

**Comment by frangio** [comment]
I've added a way to import the file directly into the docs page.

**Review by frangio — APPROVED** [review]
LGTM

**Pull request merged**